Add custom CSS support

joshuadwire opened the pull request:
* Allow custom CSS for self-hosted white-label users or Pro
invoiceninja.com users
* Restrictive sanitization for CSS on invoiceninja.com
* Add .body class to body element to allow styling the body
* Add client_view_css column to accounts table
* English-only translation text

hillelcoren commented:
Awesome!! This is a great feature. 

To be merged...

joshuadwire commented:
Quick rundown of the sanitize process done for invoiceninja.com users:

* The CSS is parsed and any HTML or invalid CSS is removed
* Only whitelisted properties are allowed
* All `@import` declarations are removed
* JavaScript and other non-safe protocols are removed from `background-image` and other URLs
* `body` selectors are changed to `.body` (The library I'm using doesn't support `body` as a selector.)
* The library currently doesn't support `rgba` colors, so that's a little disappointing.
* Potentially dangerous expressions like `calc`, `expression`, `behavior` are removed.

No sanitization is run for self-hosted installations, since there's no vulnerability with allowing them to put anything on their own site.

hillelcoren commented:
Thanks again!!

hillelcoren commented:
FYI... we're planning on adding additional client portal settings (ie, client logins) so I've changed ACCOUNT_CLIENT_VIEW_STYLING to ACCOUNT_CLIENT_PORTAL. 

I had two questions about the code:
- Can you explain the purpose of the code by 'Allow referencing the body element' in AccountController.php
- Are there any security concerns with having CSS.Trusted set to true

Thanks!

joshuadwire commented:
Sounds like a great idea. It will be nice to be able to let clients log in, especially if it will give them a way to remove/update credit card info (maybe a future feature?).

1. That code was rather interesting. HTML Purifier's main goal is to sanitize user HTML that's going to be injected into a page. It also has the capability to sanitize CSS, which is what I really needed. Because its goal is to sanitize HTML that will become only part of the page, the library doesn't support the `body` element. This apparently carries over to CSS, and it removes any use of the `body` selector.

  I wanted users to be able to style the body, so I added the `body` class to the `body` element (`<body class="body>`) and replaced all `body` selectors with `.body`. The regex is trying to select only `body` and not `.body`, `#body`, `something_body`, `bodysomething`, etc.

  Maybe a better comment would have been `// Change "body" to ".body" (HTMLPurifier removes "body" selectors)`.

  Right now, the `html` selector is just removed. I didn't add any code to support it, but it could be supported by giving it a `html` class and duplicating the code for the body.

2. From what I saw in the HTMLPurifier code, the CSS.Trusted option just allows the user to set `position`, `top`, `left`, `bottom`, `right`, and `z-index`. See https://github.com/ezyang/htmlpurifier/blob/ae1828d955112356f7677c465f94f7deb7d27a40/library/HTMLPurifier/CSSDefinition.php#L327-L329 and https://github.com/ezyang/htmlpurifier/blob/ae1828d955112356f7677c465f94f7deb7d27a40/library/HTMLPurifier/CSSDefinition.php#L409-L433.

  Most of HTML Purifier's options are designed to prevent user styles from affecting more of the page than they're supposed to. I really just wanted to prevent cross-site scripting, so I allowed pretty much all of the properties and just let HTML Purifier remove anything that could potentially load JavaScript. HTML Purifier has docs at http://htmlpurifier.org/live/configdoc/plain.html#CSS, but it's not always clear on what options do.

hillelcoren commented:
Thanks for your answers, understood.
